remove metrics proxy

pohly opened the pull request:
#### What type of PR is this?
/kind cleanup
/kind failing-test

#### What this PR does / why we need it:

The previous approach with grabbing via a nginx proxy had some
drawbacks:
- it did not work when the pods only listened on localhost (as
  configured by kubeadm) and the proxy got deployed on a different
  node
- starting the proxy raced with starting the pods, causing
  sporadic test failures because the proxy was not set up
  properly unless it saw all pods when starting the e2e.test
- the proxy was always started, whether it is needed or not
- the proxy was left running after a test and then the next
  test run triggered potentially confusing messages when
  it failed to create objects for the proxy
    
The new approach is similar to "kubectl port-forward" + "kubectl get
--raw". It uses the port forwarding feature to establish a TCP
connection via a custom dialer, then lets client-go handle TLS and
credentials.

https://github.com/kubernetes/kubernetes/pull/101895 moved the metrics
retrieval into the "do once" function callback, which had the
unintended effect that retrieving metrics data more than once no
longer worked.

#### Special notes for your reviewer:

Somehow verifying the server certificate did not work. As this
shouldn't be a big concern for E2E testing, certificate checking gets
disabled on the client side instead of investigating this further.

This replaces https://github.com/kubernetes/kubernetes/pull/101960

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

pohly committed:
e2e: grab controller and scheduler metrics via port forwarding

The previous approach with grabbing via a nginx proxy had some
drawbacks:
- it did not work when the pods only listened on localhost (as
  configured by kubeadm) and the proxy got deployed on a different
  node
- starting the proxy raced with starting the pods, causing
  sporadic test failures because the proxy was not set up
  properly unless it saw all pods when starting the e2e.test
- the proxy was always started, whether it is needed or not
- the proxy was left running after a test and then the next
  test run triggered potentially confusing messages when
  it failed to create objects for the proxy

The new approach is similar to "kubectl port-forward" + "kubectl get
--raw". It uses the port forwarding feature to establish a TCP
connection via a custom dialer, then lets client-go handle TLS and
credentials.

Somehow verifying the server certificate did not work. As this
shouldn't be a big concern for E2E testing, certificate checking gets
disabled on the client side instead of investigating this further.

pohly commented:
/cc @knight42 @aojea

pohly commented on test/e2e/suites.go:
```diff
@@ -61,7 +61,7 @@ func gatherTestSuiteMetrics() error {
 
 	// Grab metrics for apiserver, scheduler, controller-manager, kubelet (for non-kubemark case) and cluster autoscaler (optionally).
-	grabber, err := e2emetrics.NewMetricsGrabber(c, nil, !framework.ProviderIs("kubemark"), true, true, true, framework.TestContext.IncludeClusterAutoscalerMetrics)
+	grabber, err := e2emetrics.NewMetricsGrabber(c, nil, nil /* TODO */, !framework.ProviderIs("kubemark"), true, true, true, framework.TestContext.IncludeClusterAutoscalerMetrics)
```
This slipped through...

/hold

pohly replied:
/hold cancel

Fixed.

pohly commented on test/e2e/framework/metrics/metrics_grabber.go:
```diff
@@ -53,26 +54,31 @@ type Collection struct {
 	kubeScheduler                     string
-	waitForSchedulerReadyOnce         sync.Once
 	kubeControllerManager             string
-	waitForControllerManagerReadyOnce sync.Once
```
I removed the `sync.Once` check because:
- in practice it was always done because the Grabber instances only seem to be used once (otherwise https://github.com/kubernetes/kubernetes/commit/c496b1d335862cef202753d815323ce884dac862 would have broken tests)
- if called more than once, checking for readiness repeatedly shouldn't cause much overhead (pods already ready) and may be useful (in the unlikely case that pods restart)

aojea commented on test/e2e/framework/metrics/metrics_grabber.go:
```diff
@@ -300,3 +280,68 @@ func (g *Grabber) getMetricsFromPod(client clientset.Interface, podName string,
+
+// getSecureMetricsFromPodRetry is similar to getSecureMetricsFromPod except that it
+// first waits for the pod to be ready and then also retries metrics data retrieval.
+func (g *Grabber) getSecureMetricsFromPodRetry(podName string, namespace string, port int) (string, error) {
```
this `port` is not used inside thid method , is it meant to be used as `kubeControllerManagerPort`?

pohly replied:
Right, that was a cut-and-paste error. Fixed.

aojea commented on test/e2e/framework/pod/dial.go:
```diff
@@ -0,0 +1,206 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+Copyright 2021 Intel Corporation.
```
not much into licensing, but I can't see any other file with a copyright that is no "The kubernete authors"

pohly replied:
I can change that.

aojea commented on test/e2e/framework/pod/dial.go:
```diff
@@ -0,0 +1,206 @@
+	if err != nil {
+		return nil, fmt.Errorf("dialer failed: %v", err)
+	}
+	requestID := "1"
```
what are the implication of this? only one connection per pod?

pohly replied:
No, only one request per connection.

Theoretically the same connection to the API server could handle multiple different target pods, but in practice in some other context I found it more stable to not use that.

aojea commented on test/e2e/framework/pod/dial.go:
```diff
@@ -0,0 +1,206 @@
+	if err != nil {
+		return nil, fmt.Errorf("error creating error stream: %v", err)
+	}
+	errorStream.Close()
```
I had a comment here asking if this should be a defer() or do we want to close it immediately after creating it

pohly replied:
This was copied from https://github.com/kubernetes/kubernetes/blob/1d38084930b56be9b5830c20bd5b464b1f955047/staging/src/k8s.io/client-go/tools/portforward/portforward.go#L338-L344

My understanding is that this is a two-way stream and closing the local writer side is fine.

pacoxu replied:
Keep the comment ` // we're not writing to this stream ` here may help us understand the code.

pohly replied:
It's still there:
https://github.com/kubernetes/kubernetes/blob/7930929678a933f766b0b3837a1fcdd80692f103/test/e2e/framework/pod/dial.go#L111

I just moved it up a bit and added additional comments because the original code did not explain what the stream was needed for.

aojea commented on test/e2e/framework/pod/dial.go:
```diff
@@ -0,0 +1,206 @@
+// by Addr.String.
+func ParseAddr(addr string) (*Addr, error) {
+	parts := addrRegex.FindStringSubmatch(addr)
+	if parts == nil {
```
`!! len(parts) != 3` ?

aojea replied:
ah, `FindStringSubmatch` regexp seems to guarantee that

pohly commented:
Getting there... I checked that the boilerplate test passes by running it locally, so the next `pull-kubernetes-verify` should be okay.

pohly commented:
/retest

knight42 commented:
@pohly @aojea It seems the error reported in job `pull-kubernetes-e2e-gce-ubuntu-containerd` is valid:

```
 "error waiting for pod kube-system/kube-scheduler-e2e-3874a48290-a7d53-master to expose metrics: timed out waiting for the condition; Get \"https://kube-system.kube-scheduler-e2e-3874a48290-a7d53-master:10259/metrics?timeout=32s\": dialer failed: unable to upgrade connection: Debug endpoints are disabled.",
```

Now debugging endpoints are actually disabled on master nodes: https://github.com/kubernetes/kubernetes/issues/97383#issuecomment-749901525. 

If we want to port-forward to the pods(the scheduler and controller-manager pod) on master nodes, we have to enable the debugging endpoints.

aojea commented:
> If we want to port-forward to the pods(the scheduler and controller-manager pod) on master nodes, we have to enable the debugging endpoints.

:grimacing: can this be more difficult ? :upside_down_face:

knight42 commented:
> 😬 can this be more difficult ? 🙃

@aojea I think that would be pretty easy, just change this line https://github.com/kubernetes/kubernetes/blob/382a33986b043f78a42d3d865d0ca383687bf171/cluster/gce/util.sh#L1028 to
```
enableDebuggingHandlers: true
```
but I dont know whether it is safe to do that.

aojea commented:
> but I dont know whether it is safe to do that.

I hit this before, I think that debugging handlers are disabled in all control-plane nodes on purpose , @BenTheElder or @spiffxp will know better

pohly commented:
So the API endpoint that is used by port-forwarding is considered an optional debug feature?

pohly commented:
> So the API endpoint that is used by port-forwarding is considered an optional debug feature?

Sure looks like it:
https://github.com/pohly/kubernetes/blob/fix/deflake-metrics-proxy/pkg/kubelet/server/server.go#L497-L517

:cry: 

Not sure how to proceed here.

aojea commented:
> Not sure how to proceed here.

Host exec curl and process the output?

liggitt commented:
> > Not sure how to proceed here.
> 
> Host exec curl and process the output?

is this used by conformance tests? if so, host execing isn't permitted, right (per https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md#conformance-test-requirements)?

aojea commented:
> is this used by conformance tests? if so, host execing isn't permitted, right (per https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md#conformance-test-requirements)?

on windows, you have to tag them as [LinuxOnly] 
> Networking
> Pods set HostNetwork=true. Is not supported on Windows, and the Pod will not start.

pohly commented:
> Host exec curl and process the output?

That has the exact same problem, exec is also one of those debug handlers that are disabled.

> is this used by conformance tests?

No. That's why I thought that relying on port-forwarding would be okay. I already knew that it is not required for conformance, although the exact reason for that wasn't made clear when I asked about it on Slack.

I suppose there's no way to query whether the cluster is configured to support port-forwarding for the master node. We could try the operation and then skip the test if that particular problem is detected. But both defeats the purpose of the test to some extend - what if the feature was meant to be turned on and is only off by accident?

pohly commented:
We could run a pod which executes curl. But the overhead would be high if done frequently. OTOH, I don't think kube-controller-manager and kube-scheduler metrics are queried often.

aojea commented:
more crazy ideas, exposing the metrics port using a service?

pohly commented:
> We could run a pod which executes curl.

But how can we then get the log output of that pod? `containerLogs` is also disabled...

pohly commented:
How do you connect to a service from outside the cluster? Doesn't that again depend on port-forwarding to be portable?

pohly commented:
Create a service, then run a container with curl on some *other* node where debug handlers are not disabled... definitely getting fairly crazy.

Perhaps the approach with nginx is simpler after all? But then the remaining issues with it should be fixed (like running the proxy on the master node).

pohly commented:
> Perhaps the approach with nginx is simpler after all? But then the remaining issues with it should be fixed (like running the proxy on the master node).

... and once it runs there, connecting to nginx through the APIserver will also not work.

pohly commented:
Or at least that is my understanding of these debug handlers. I don't know how it worked for the non-secure port before - those were reached through the api server.

fedebongio commented:
/cc @logicalhan

logicalhan commented:
/assign
/triage accepted

logicalhan commented:
/cc @pacoxu

pacoxu commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd
The flake is not solved yet.

pacoxu commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd
The flake is not solved yet.

pohly commented:
> The flake is not solved yet.

It's not a flake. Port-forwarding, a feature that this PR depends on, is not enabled in those clusters on the master node. It's not clear what approach should be used instead to get the metrics data.

After thinking about this some more, I'm coming to the conclusion that these clusters should be reconfigured to allow port-forwarding. My rationale is that the jobs are used to run more than just conformance tests and therefore can and should configured so that these additional tests can work if rewriting tests to not depend on optional features is hard.

I believe that applies here - I could try out some of the alternative solutions that I outlined above, but IMHO all of them will be more complex, if they work at all.

pohly committed:
cluster: enable debug handlers on GCE master nodes

This is needed for testing metrics support via the secure port
of kube-scheduler and kube-controller-manager. To access that
port, port-forwarding is used.

pohly commented:
I have added a commit with `enableDebuggingHandlers: true`. Let's see whether that helps and then discuss further whether we want to make that change.

pohly commented:
/retest

logicalhan commented:
Would you mind rebasing this? I'm curious if this will fix the test.

BenTheElder commented on cluster/gce/util.sh:
```diff
@@ -1025,7 +1025,11 @@ EOF
+# kube-controller-manager and kube-scheduler
+# when retrieving metrics data
+# (see test/e2e/framework/metrics/metrics_grabber.go).
+enableDebuggingHandlers: true
```
this should only be done in the test config, not the non-test config.

pohly replied:
Which "test config"? Do you mean some Prow job definition?

pacoxu replied:
I think it means `cluster/gce/config-test.sh`.
Take `KUBELET_TEST_LOG_LEVEL` as an exmaple, and use it here.

pohly replied:
So don't hard-code in `util.sh` but rather make it configurable in `config-test.sh`. I added `KUBELET_ENABLE_DEBUGGING_HANDLERS` (default true) and `MASTER_KUBELET_ENABLE_DEBUGGING_HANDLERS` (default `KUBELET_ENABLE_DEBUGGING_HANDLERS` and thus true).

This is still a change in how the cluster is created by default, though. If that is not desired, then we need to change Prow jobs.

BenTheElder replied:
> This is still a change in how the cluster is created by default, though. If that is not desired, then we need to change Prow jobs.

Yes the default cluster without config-test.sh should not be enabling these without opt-in. CI should be using config-test?

BenTheElder commented:
> After thinking about this some more, I'm coming to the conclusion that these clusters should be reconfigured to allow port-forwarding. My rationale is that the jobs are used to run more than just conformance tests and therefore can and should be configured so that these additional tests can work if rewriting tests to not depend on optional features is hard.

some of the jobs are, but we're also generally trying to make tests plausible for conformance.

these tests are also going to be a bit less useful if they require enabling a pretty obvious security hazard that doesn't have much legitimate use in a production cluster 😬

pohly commented:
> these tests are also going to be a bit less useful if they require enabling a pretty obvious security hazard that doesn't have much legitimate use in a production cluster

The tests that are still failing are about retrieving metrics data from the control plane from outside of a cluster - something that most likely will be tightly controlled and/or disabled in a production cluster. So it's not surprising that they don't work unless the cluster is prepared to allow that.

That the tests are enabled by default probably comes from the desire to ensure that the test code itself works. They are not about testing behavior of the cluster. From what I remember, the code then is just used to collect optional information about a test run, for example in test/e2e/e2e.go. But I might be wrong, I only got involved recently :shrug:

pohly commented:
> Would you mind rebasing this? I'm curious if this will fix the test.

`MetricsGrabber should grab all metrics from API server` already worked before the conflict in https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/102050/pull-kubernetes-e2e-gce-ubuntu-containerd/1395341056744099840/

I rebased, just in case.

pohly commented:
/retest

For "pull-kubernetes-e2e-gce-storage-slow — Pod got deleted unexpectedly"

aojea commented:
what definitively has to go away is to run always the pod once the e2e framework starts, on cluster that doesn't provide the controller-manager and the scheduler as pods it blocks the e2e test

```
un  2 15:54:57.709: INFO: kube-apiserver version: v1.21.1-1332+4b2b6ff4947ea1-dirty
I0602 15:54:57.833105  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:54:57.833138  499681 metrics_proxy.go:73] Found components: []
I0602 15:55:03.076471  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:55:03.076506  499681 metrics_proxy.go:73] Found components: []
I0602 15:55:08.076374  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:55:08.076408  499681 metrics_proxy.go:73] Found components: []
I0602 15:55:13.072322  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:55:13.072357  499681 metrics_proxy.go:73] Found components: []
I0602 15:55:18.074051  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:55:18.074095  499681 metrics_proxy.go:73] Found components: []
I0602 15:55:23.075844  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:55:23.075900  499681 metrics_proxy.go:73] Found components: []
I0602 15:55:28.073385  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:55:28.073421  499681 metrics_proxy.go:73] Found components: []
I0602 15:55:33.071707  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:55:33.071743  499681 metrics_proxy.go:73] Found components: []
I0602 15:55:38.070367  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:55:38.070406  499681 metrics_proxy.go:73] Found components: []
I0602 15:55:43.071004  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:55:43.071024  499681 metrics_proxy.go:73] Found components: []
I0602 15:55:48.072593  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:55:48.072628  499681 metrics_proxy.go:73] Found components: []
I0602 15:55:53.070785  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:55:53.070821  499681 metrics_proxy.go:73] Found components: []
I0602 15:55:58.072652  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:55:58.072690  499681 metrics_proxy.go:73] Found components: []
I0602 15:56:03.069078  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:56:03.069110  499681 metrics_proxy.go:73] Found components: []
I0602 15:56:08.071324  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:56:08.071366  499681 metrics_proxy.go:73] Found components: []
I0602 15:56:13.090852  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:56:13.092311  499681 metrics_proxy.go:73] Found components: []
I0602 15:56:18.071840  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:56:18.071878  499681 metrics_proxy.go:73] Found components: []
I0602 15:56:23.073357  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:56:23.073393  499681 metrics_proxy.go:73] Found components: []
I0602 15:56:28.075856  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:56:28.075897  499681 metrics_proxy.go:73] Found components: []
I0602 15:56:33.070714  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:56:33.070758  499681 metrics_proxy.go:73] Found components: []
I0602 15:56:38.072853  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:56:38.072889  499681 metrics_proxy.go:73] Found components: []
I0602 15:56:43.072344  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:56:43.072383  499681 metrics_proxy.go:73] Found components: []
I0602 15:56:48.072135  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:56:48.074280  499681 metrics_proxy.go:73] Found components: []
I0602 15:56:53.069409  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:56:53.069447  499681 metrics_proxy.go:73] Found components: []
I0602 15:56:58.074682  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:56:58.074717  499681 metrics_proxy.go:73] Found components: []
I0602 15:57:03.076520  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:57:03.076551  499681 metrics_proxy.go:73] Found components: []
I0602 15:57:08.072875  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:57:08.072913  499681 metrics_proxy.go:73] Found components: []
I0602 15:57:13.073073  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:57:13.073135  499681 metrics_proxy.go:73] Found components: []
I0602 15:57:18.072659  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:57:18.072702  499681 metrics_proxy.go:73] Found components: []
I0602 15:57:23.074485  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:57:23.074522  499681 metrics_proxy.go:73] Found components: []
I0602 15:57:28.071219  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:57:28.071251  499681 metrics_proxy.go:73] Found components: []
I0602 15:57:33.075648  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:57:33.075687  499681 metrics_proxy.go:73] Found components: []
I0602 15:57:38.071207  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:57:38.071239  499681 metrics_proxy.go:73] Found components: []
I0602 15:57:43.076463  499681 metrics_proxy.go:72] Only 0 components found. Will retry.
I0602 15:57:43.076503  499681 metrics_proxy.go:73] Found components: []
```

liggitt commented:
> what definitively has to go away is to run always the pod once the e2e framework starts, on a cluster that doesn't provide the controller-manager and the scheduler as pods it blocks the e2e test

At this point, my instinct is either 1) to revert the change that required the metrics proxy to get back to the previous working state or 2) remove the metrics bit (since it is apparently optional to begin with) and regroup about what we want these e2e tests to gather.

Continuing in master with flaky tests for weeks is not a good way to approach this.

knight42 commented:
> 1) to revert the change that required the metrics proxy to get back to the previous working state

@liggitt #101960 is intended to deflake the tests that need to grab metrics from controller-manager or scheduler, the tests in master should be less flaky now.

logicalhan commented:
> At this point, my instinct is either 1) to revert the change that required the metrics proxy to get back to the previous working state or 2) remove the metrics bit (since it is apparently optional to begin with) and regroup about what we want these e2e tests to gather.

If it is optional then we should do 2. Let's figure out what to do without simultaneously destroying our CI pipeline.

pohly commented:
> If it is optional then we should do 2.

Well, not quite optional. There are tests that use metrics gathering, if available, to test expected behavior of core components. This seems like useful testing to have, even if it is not guaranteed that it can be used for a cluster.

Example: https://github.com/kubernetes/kubernetes/commit/564e531aa72f48afde659e5ce464d0c35d608889#diff-74a1bbe84c67d45f3d4e437f9dcfc98bcae740c9a88e05b803ef284f5a1f6ff0

What about reconfiguring the gce clusters to allow retrieval of metrics data as proposed in this PR? Is that not an option?

pohly commented:
> 1) to revert the change that required the metrics proxy to get back to the previous working state

Isn't that change the disabling of the insecure port? That probably shouldn't be reverted. Allowing debug handlers on the master node seems to be the lesser evil.

liggitt commented:
> Well, not quite optional. There are tests that use metrics gathering, if available, to test expected behavior of core components.

"using if available" seems like the definition of optional :)

aojea commented:
I'm trying to make this pod proxy setup optional, https://github.com/kubernetes/kubernetes/pull/102535 , so it is only invoked when necessary.

pohly commented:
> "using if available" seems like the definition of optional :)

Yes, but "remove the metrics bits" is stronger than merely making the tests optional. Without the metrics bits, the tests have to be removed altogether.

pohly commented:
We could check in advance whether debug handlers are enabled for those pods. There doesn't seem to be a good way to do that other than trying some operation that needs them, but it would achieve the goal of automatically enabling the metrics retrieval only when it actually works. Then tests can skip the testing if that is not the case.

Existence of the pods is already checked that way.

pohly committed:
e2e metrics: check whether debug handlers are available

This can be checked by trying to retrieve log output. As in the case
of no pod found, a warning gets emitted when log retrieval fails and
metrics grabbing gets disabled.

Logging is checked instead of actual metrics retrieval because the
latter is more complex and thus more likely to fail for other reasons.

pohly commented:
> We could check in advance whether debug handlers are enabled for those pods.

I've implemented that and removed the gce cluster config change.

@aojea I still prefer port forwarding over the proxy approach, even if the proxy gets started better (i.e. https://github.com/kubernetes/kubernetes/pull/102535). The advantages are that the cluster does not get modified permanently and that it works for kubeadm clusters.

aojea commented:
> @aojea I still prefer port forwarding over the proxy approach, even if the proxy gets started better (i.e. #102535). The advantages are that the cluster does not get modified permanently and that it works for kubeadm clusters.

just opening new avenues and checking feasibility so we can take an informed decision :)

pohly commented:
Looks like my assumptions about debug handlers are not true. Log retrieval worked, but then port forwarding didn't. This needs further work. But before I invest more time, can someone comment whether this approach then would be acceptable?

BenTheElder commented:
> Yes, but "remove the metrics bits" is stronger than merely making the tests optional. Without the metrics bits, the tests have to be removed altogether.

Optional tests are generally behind a `[Feature: Foo]` so that we don't have:
- not running when we thought they were
- running and failing when a cluster does not support them

And IMHO are of pretty low value in the general case, and of some value in the uncommon case when a SIG closely monitors a special job to run them.

pohly commented:
> Optional tests are generally behind a [Feature: Foo]

Isn't that used for alpha features that are not meant to run in "normal" jobs? Suppose a SIG has a test that can run fine in the current `pull-kubernetes-e2e-kind` and uses metrics. Do we now need a `pull-kubernetes-e2e-kind-metrics` job for running tests tagged with `[Feature: Metrics]`? Or can the job config of `pull-kubernetes-e2e-kind` be changed to enable tests with a `[Feature: Metrics]` tag?

pohly committed:
e2e metrics: skip tests when metrics grabbing is disabled

The MetricsGrabber checked whether a component supported metrics
grabbing, but then tests didn't have an API to use the result of that
check. Because metrics grabbing is an optional debug feature, tests
must skip checks that depend on metrics data or, when the entire
test is about metrics data, skip the test.

This is now supported with a special error that gets wrapped and
returned by the individual Grab functions.

pohly commented:
This whole `MetricsGrabber` API is... in need of some considerable redesign. My code for detecting whether metrics grabbing works (both approaches!), but the tests then didn't (and couldn't) use that information because the API doesn't expose it and therefore the tests just failed.

I've now tested on a local cluster with and without debug handlers and got the tests to skip properly if (and only if) debug handlers were disabled.

pohly commented:
[pull-kubernetes-e2e-gce-ubuntu-containerd](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/102050/pull-kubernetes-e2e-gce-ubuntu-containerd/1400372705852657664/):

"MetricsGrabber should grab all metrics from a ControllerManager" and "from Scheduler" got skipped because debug handlers are disabled.

[pull-kubernetes-e2e-kind](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/102050/pull-kubernetes-e2e-kind/1400372705898795008):

They get skipped because of https://github.com/kubernetes/kubernetes/blob/2bbc0e3c54fdf11b0006efe7f65f6a416ea15d32/test/e2e/instrumentation/monitoring/metrics_grabber.go#L83-L86

I don't know what that is about :shrug:  Probably doesn't apply anymore. Let's try without that check...

pohly committed:
e2e metrics: remove redundant checks around metrics tests

The MetricsGrabber itself knows now whether it supports each
component. The checks inside the tests therefore are redundant at best
or worse, they are wrong: for example, on a KinD cluster the check for
"has master node registered" failed and metrics grabbing from
scheduler and controller manager were skipped unnecessarily.

aojea commented on test/e2e/framework/metrics/metrics_grabber.go:
```diff
@@ -102,37 +123,48 @@ func NewMetricsGrabber(c clientset.Interface, ec clientset.Interface, kubelets b
 	}, nil
 }
 
+func checkPodDebugHandlers(c clientset.Interface, requested bool, component, podName string) bool {
```
is not a bit weird to have `requested` here? if the caller knows he is not going to request info why he calls the function anyway?
I see that it simplifies the code in `NewMetricsGrabber` constructor but is it worth it?

pohly replied:
> I see that it simplifies the code in NewMetricsGrabber constructor but is it worth it?

This is an internal helper function. It's whole purpose is to make the constructor simpler and not have any repeated code there (like this if check), so I think it's worth it.

pohly replied:
> if the caller knows he is not going to request info why he calls the function anyway?

The caller is the constructor and it doesn't know whether metrics for a component got requested. It's repeating that if check in the constructor that gets avoided by moving it into a parameter.

aojea commented on test/e2e/framework/metrics/metrics_grabber.go:
```diff
@@ -102,37 +123,48 @@ func NewMetricsGrabber(c clientset.Interface, ec clientset.Interface, kubelets b
+	// The debug handlers on the host where the pod runs might be disabled.
+	// We can check that indirectly by trying to retrieve log output.
+	limit := int64(1)
+	if _, err := c.CoreV1().Pods(metav1.NamespaceSystem).GetLogs(podName, &v1.PodLogOptions{LimitBytes: &limit}).DoRaw(context.TODO()); err != nil {
```
is there any possible race here that the pod is listed, handlers are available but we fail to get the logs?
is there an specific error we can match against?

pohly replied:
This could fail because the containers aren't running yet (IMHO unlikely, because this is now getting called after the e2e.test binary has already ensured that the cluster is up) or simply because of load.

There is some error text about debug handlers that we could check for and we could retry for other errors. I decided against it because its more complex, unlikely to happen (admittedly a bit subjective) and the impact of incorrectly disabling metrics grabbing is low. Run a job often enough and most of the time it should execute the tests.

jingxu97 commented:
I noticed that pdcsi driver migration test for windows start failing after #101895 is merged. The error is 

```
/tmp/gcp-pd-driver-tmp430967419/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/testsuites/volumes.go:196
W0604 14:33:43.314666   85793 metrics_grabber.go:118] Did not receive an external client interface. Grabbing metrics from ClusterAutoscaler is disabled.
Jun  4 14:35:47.479: FAIL: Error getting c-m metrics : error waiting for controller manager pod to expose metrics: timed out waiting for the condition; the server is currently unable to handle the request (get pods metrics-proxy:10257)
```
But I could not tell why in-tree test can pass but csi migration for Windows failed (Linux tests are fine)
@pohly do you think this PR will help fix the issue?

pohly commented:
> I noticed that pdcsi driver migration test for windows start failing after #101895 is merged. The error is
> 
> ```
> /tmp/gcp-pd-driver-tmp430967419/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/testsuites/volumes.go:196
> W0604 14:33:43.314666   85793 metrics_grabber.go:118] Did not receive an external client interface. Grabbing metrics from ClusterAutoscaler is disabled.
> Jun  4 14:35:47.479: FAIL: Error getting c-m metrics : error waiting for controller manager pod to expose metrics: timed out waiting for the condition; the server is currently unable to handle the request (get pods metrics-proxy:10257)
> ```
> 
> But I could not tell why in-tree test can pass but csi migration for Windows failed (Linux tests are fine)

The current solution in master is known to fail when kube-controller-manager only listens on localhost. That is the case when setting up a cluster with kubeadm. Is the Windows job perhaps running into that?

> @pohly do you think this PR will help fix the issue?

If it is the reason mention above, then yes.

jingxu97 commented:
> setting up a cluster with kubeadm

Windows test job config is here https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/gcp-compute-persistent-disk-csi-driver-windows.yaml#L226

I don't think it is using kubeadm to set up cluster?

In this case, it that possible because in Windows cluster, proxy pod is not yet ready when trying to grab metrics?

pohly commented:
> > setting up a cluster with kubeadm
> 
> Windows test job config is here https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/gcp-compute-persistent-disk-csi-driver-windows.yaml#L226
> 
> I don't think it is using kubeadm to set up cluster?

I don't know. In such a running cluster, what is the `--bind-address` parameter of kube-controller-manager? With kubeadm, it is `--bind-address=127.0.0.1` and the proxy cannot connect.

> In this case, it that possible because in Windows cluster, proxy pod is not yet ready when trying to grab metrics?

No, the code should wait for the proxy to be ready:
https://github.com/kubernetes/kubernetes/blob/028bffba588c8cd4685e783dbab768a2ee15c824/test/e2e/framework/metrics/metrics_proxy.go#L208

pohly commented:
/retest

BenTheElder commented:
> Isn't that used for alpha features that are not meant to run in "normal" jobs? 

`[Feature:Foo]` is used to tag things that cannot reasonably be expected to be enabled in all clusters. E.g. say a test for hugepages. It's not about the jobs, it's about people running the tests and opting into running tests that require special functionality that is not standard (e.g. re-enabling debug handlers...)

> Suppose a SIG has a test that can run fine in the current pull-kubernetes-e2e-kind and uses metrics. Do we now need a pull-kubernetes-e2e-kind-metrics job for running tests tagged with [Feature: Metrics]? Or can the job config of pull-kubernetes-e2e-kind be changed to enable tests with a [Feature: Metrics] tag?

My brain is fuzzy right now but I think we can tweak the job to include this, but also if it's going to pass on a reasonable default cluster config we don't need the tag probably.

aojea commented:
lgtm
pulling Jordan since he was in the original thread 
/cc @liggitt

pohly commented:
> My brain is fuzzy right now but I think we can tweak the job to include this, but also if it's going to pass on a reasonable default cluster config we don't need the tag probably.

I think with the current state of the PR we do not need a feature tag. The metrics manager now auto-detects whether debug handlers are enabled. Tests that use it need to skip some part of the test or the entire test.

We could still add a feature tag and set it in jobs which are supposed to have debug handlers, but IMHO that's overkill.

BenTheElder commented:
SGTM -- especially if we're not trying to make any of these tests conformance.

Thanks for working on this.

aojea commented:
/lgtm

BenTheElder reviewed (commented):
Are we now skipping all of these tests in CI with this change?

I recall previously in an earlier iteration there was an attempt at altering at least kube-up config

aojea commented:
/hold
/lgtm cancel
this is not working correctly

kind job - test pass , but is not clear to me from the logs
````
BeforeEach] [sig-instrumentation] MetricsGrabber
  /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:185
[1mSTEP[0m: Creating a kubernetes client
Jun 16 10:35:48.120: INFO: >>> kubeConfig: /root/.kube/kind-test-config
[1mSTEP[0m: Building a namespace api object, basename metrics-grabber
[1mSTEP[0m: Waiting for a default service account to be provisioned in namespace
[BeforeEach] [sig-instrumentation] MetricsGrabber
  /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/instrumentation/monitoring/metrics_grabber.go:39
W0616 10:35:48.169641   89550 metrics_grabber.go:127] Did not receive an external client interface. Grabbing metrics from ClusterAutoscaler is disabled.
W0616 10:35:48.190371   89550 metrics_grabber.go:151] Can't find snapshot-controller pod. Grabbing metrics from snapshot-controller is disabled.
[It] should grab all metrics from a Kubelet.
  /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/instrumentation/monitoring/metrics_grabber.go:62
[1mSTEP[0m: Proxying to Node through the API server
[AfterEach] [sig-instrumentation] MetricsGrabber
  /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:186
Jun 16 10:35:48.903: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
[1mSTEP[0m: Destroying namespace "metrics-grabber-8274" for this suite.

[32m•[0m
[90m------------------------------[0m
{"msg":"PASSED [sig-instrumentation] MetricsGrabber should grab all metrics from a Kubelet.","total":-1,"completed":25,"skipped":235,"failed":0}
```

gce job - test pass, but it can't debug handlers are disabled
```
I0622 19:33:33.442] [1mSTEP[0m: Waiting for a default service account to be provisioned in namespace
I0622 19:33:33.442] [BeforeEach] [sig-instrumentation] MetricsGrabber
I0622 19:33:33.443]   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/instrumentation/monitoring/metrics_grabber.go:39
I0622 19:33:33.443] W0622 19:33:28.269293  102110 metrics_grabber.go:127] Did not receive an external client interface. Grabbing metrics from ClusterAutoscaler is disabled.
I0622 19:33:33.443] W0622 19:33:28.742104  102110 metrics_grabber.go:159] Can't retrieve log output of kube-controller-manager-e2e-75366d2e98-a7d53-master ("the server does not allow this method on the requested resource (get pods kube-controller-manager-e2e-75366d2e98-a7d53-master)"). Debug handlers might be disabled in kubelet. Grabbing metrics from kube-controller-manager is disabled.
I0622 19:33:33.444] W0622 19:33:29.203858  102110 metrics_grabber.go:159] Can't retrieve log output of kube-scheduler-e2e-75366d2e98-a7d53-master ("the server does not allow this method on the requested resource (get pods kube-scheduler-e2e-75366d2e98-a7d53-master)"). Debug handlers might be disabled in kubelet. Grabbing metrics from kube-scheduler is disabled.
I0622 19:33:33.444] [It] should grab all metrics from a Kubelet.
I0622 19:33:33.444]   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/instrumentation/monitoring/metrics_grabber.go:62
I0622 19:33:33.444] [1mSTEP[0m: Proxying to Node through the API server
I0622 19:33:33.445] [AfterEach] [sig-instrumentation] MetricsGrabber
I0622 19:33:33.445]   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:186
I0622 19:33:33.445] Jun 22 19:33:32.657: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
I0622 19:33:33.445] [1mSTEP[0m: Destroying namespace "metrics-grabber-1990" for this suite.
I0622 19:33:33.445] 
I0622 19:33:33.445] 
I0622 19:33:33.445] [32mâ€¢ [SLOW TEST:7.830 seconds][0m
I0622 19:33:33.446] [sig-instrumentation] MetricsGrabber
I0622 19:33:33.446] [90m/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/instrumentation/common/framework.go:23[0m
I0622 19:33:33.446]   should grab all metrics from a Kubelet.
I0622 19:33:33.446]   [90m/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/instrumentation/monitoring/metrics_grabber.go:62[0m
I0622 19:33:33.446] [90m------------------------------[0m
I0622 19:33:33.447] {"msg":"PASSED [sig-instrumentation] MetricsGrabber should grab all metrics from a Kubelet.","total":-1,"completed":4,"skipped":19,"failed":0}
```

aojea commented on test/e2e/instrumentation/monitoring/metrics_grabber.go:
```diff
@@ -17,61 +17,54 @@ limitations under the License.
+			e2eskipper.Skipf("%v", err)
+		}
 		framework.ExpectNoError(err)
 		gomega.Expect(response).NotTo(gomega.BeEmpty())
```
can we log the response in all tests?
 currently is not clear from the logs if the test passes or not

pohly replied:
Where do you want to add logging? In the caller of `e2eskipper.Skipf` or in `Skipf` itself?

Not being able to determine that and why a test was skipped looks like a general problem. I remember that JUnit records the skip reason, but prow.k8s.io doesn't show that. Therefore I am leaning towards logging in `Skipf` as a band-aid.

aojea replied:
I was thinking just to log the whole response
```
framework.Logf("Returned metrics: %v", response)
```
The skip can have a more complete message as in:
```
e2eskipper.Skipf("Skipping test, can't grab metrics: %v", err)
```

BenTheElder replied:
>  I remember that JUnit records the skip reason, but prow.k8s.io doesn't show that.

If there is junit data that is not possible to view on prow.k8s.io please file a test-infra issue so we can see if someone can fix this.

pohly replied:
> I was thinking just to log the whole response

Isn't that going to be a bit verbose? But I can add it.

> The skip can have a more complete message as in:
> `e2eskipper.Skipf("Skipping test, can't grab metrics: %v", err)`

That makes less sense to me:
- It's a skip message, explicitly adding "Skipping test" is redundant. Other calls to `Skipf` don't have it.
- `can't grab metrics` is redundant because the error already contains `metrics grabbing disabled`

> If there is junit data that is not possible to view on prow.k8s.io please file a test-infra issue so we can see if someone can fix this.

The JUnit file must have been from some other test tool, the junit files from e2e.test don't contain the skip reason. That makes it an issue for k/k: https://github.com/kubernetes/kubernetes/issues/103430

Whether prow.io then would show the skip reason remains to be seen.

aojea replied:
feel free to scratch those comments, they are very hacky

aojea replied:
hmm I think that the problem is that the test is marked as PASSED instead of SKIPPED

aojea replied:
sorry, last day before holidays, I'm not thinking straight today, it does show as skipped

BenTheElder replied:
It should be marked skipped when skipped, if it's not marked skipped it could only be that some conditional was used internally. I don't think we need to log skipping specially here but it would certainly be nice if we recorded why tests were skipped more generally

aojea commented:
> Are we now skipping all of these tests in CI with this change?
> 
> I recall previously in an earlier iteration there was an attempt at altering at least kube-up config

@pohly we need to enable the debugging handlers, can you incorporate the change so we don't loose coverage in CI?

> it is a badly named flag, but it is core function and should remain on.


xref https://github.com/kubernetes/kubernetes/pull/64515#issuecomment-393326208

```patch
diff --git a/cluster/gce/util.sh b/cluster/gce/util.sh
index 4e2e48cda2f..42380ea51e9 100755
--- a/cluster/gce/util.sh
+++ b/cluster/gce/util.sh
@@ -1025,7 +1025,7 @@ EOF
 # cat the Kubelet config yaml for masters
 function print-master-kubelet-config {
   cat <<EOF
-enableDebuggingHandlers: false
+enableDebuggingHandlers: true
 hairpinMode: none
 staticPodPath: /etc/kubernetes/manifests
 authentication:
```

pohly commented:
> we need to enable the debugging handlers, can you incorporate the change so we don't loose coverage in CI?

I thought people were against changing cluster configurations and instead wanted the tests to adapt to how the cluster is configured?

Regarding test coverage, pull-kubernetes-e2e-kind  is configured so that the "MetricsGrabber should grab all metrics" tests run (https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/102050/pull-kubernetes-e2e-kind/1405103662593740800).

If we want to be *sure* that they run and not just get skipped we have to:
- add a new  feature tag `MetricsGrabber` to pull-kubernetes-e2e-kind
- add a variant of these tests that use that tag and do not skip or
- keep one variant that only runs with the feature tag.

I prefer to have two variants, with and without feature tag, because then we get better coverage.

BenTheElder commented:
er, I think the test change is probably fine, but we should change kube-up's test-config.sh so that we have these in CI there? (since that's still most CI coverage currently)

BenTheElder commented:
I think people are only against changing cluster config when we want to potentially have conformance tests (e.g. kind config is expected to match relatively standard kubeadm and we run a job where we disable all non-GA APIs to help ensure conformance tests are reasonable).

Where we already have CI opt into special "test" cluster config, it makes sense to enable more kinds of test. My previous comment re: GCE changes was just about moving it to test-config.sh IIRC.

aojea commented:
> I thought people were against changing cluster configurations and instead wanted the tests to adapt to how the cluster is configured?

I thought the same, but this seems to change a default configuration that seems to be commonly enabled xref #64515

> : GCE changes was just about moving it to test-config.sh

ahahh, I was crazy trying to find the file, you mean config-test.sh, right? https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/config-test.sh

pohly commented:
> My previous comment re: GCE changes was just about moving it to test-config.sh 

That's how I had already implemented it back in March in 7d18b72de0f43f39b3c755715af20cc7e6350dec :grin: 

So let's restore that commit... done.

aojea commented:
it runs now in kind and gce, and shows skipped correctly the ones that are not able to get metrics like

> Kubernetes e2e suite: [sig-instrumentation] MetricsGrabber should grab all metrics from a Scheduler.

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/102050/pull-kubernetes-e2e-gce-ubuntu-containerd/1410880738974240768/

/hold cancel
/lgtm

Thanks

pohly commented:
/retest

Unrelated flakes.

pohly commented:
/hold

Let's be sure that the change in `cluster/gce/config-test.sh` which turns on debug handlers by default is okay.

BenTheElder commented on cluster/gce/config-test.sh:
```diff
@@ -236,6 +236,10 @@ TEST_CLUSTER_RESYNC_PERIOD=${TEST_CLUSTER_RESYNC_PERIOD:---min-resync-period=3m}
 
+# Enable debug handlers (port forwarding, exec, container logs, etc.).
+KUBELET_ENABLE_DEBUGGING_HANDLERS=${KUBELET_ENABLE_DEBUGGING_HANDLERS:-true}
+MASTER_KUBELET_ENABLE_DEBUGGING_HANDLERS=${MASTER_KUBELET_ENABLE_DEBUGGING_HANDLERS:-${KUBELET_ENABLE_DEBUGGING_HANDLERS}}
```
we can iterate on this later if we find some reason we don't want this to be the default, I think it makes sense for clusters with test-config

BenTheElder reviewed (approved):
/lgtm
/approve
thank you, and thank you for your patience

BenTheElder commented:
> That's how I had already implemented it back in March in 7d18b72 😁

in 7d18b72 it is on by default even without the config-test

 https://github.com/kubernetes/kubernetes/commit/7d18b72de0f43f39b3c755715af20cc7e6350dec#diff-09816a4dd2fd3cf28f97bf9f34330a66336c631830b103d0c96d71eb632dab0dL1028 

but that's since been fixed.

BenTheElder commented:
/hold cancel

pohly commented:
> I noticed that pdcsi driver migration test for windows start failing after #101895 is merged. The error is
> 
> ```
> /tmp/gcp-pd-driver-tmp430967419/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/testsuites/volumes.go:196
> W0604 14:33:43.314666   85793 metrics_grabber.go:118] Did not receive an external client interface. Grabbing metrics from ClusterAutoscaler is disabled.
> Jun  4 14:35:47.479: FAIL: Error getting c-m metrics : error waiting for controller manager pod to expose metrics: timed out waiting for the condition; the server is currently unable to handle the request (get pods metrics-proxy:10257)
> ```
> 
> But I could not tell why in-tree test can pass but csi migration for Windows failed (Linux tests are fine)
> Example run: https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-gce-pd-csi-driver-latest-k8s-master-windows-2019-migration/1400809226237906944
> 
> @pohly do you think this PR will help fix the issue?

Did you work around the metrics test failure? I don't see the "error waiting for controller manager pod to expose metrics" from the example run in any of the recent job runs, including those shortly before this PR was merged.

johngmyers commented:
@pohly could you provide a reference for the requirements for the controller manager and scheduler server certificates? I don't see them mentioned in https://kubernetes.io/docs/setup/best-practices/certificates/

It looks like this PR requires them to have "localhost" in the SAN, which means that if they were to be issued from the Kubernetes General CA, those services could impersonate kube-apiserver to other clients running on apiserver nodes. But the code in this PR ignores the cert signer, which removes the entire point of using TLS in the first place.

pohly commented:
> @pohly could you provide a reference for the requirements for the controller manager and scheduler server certificates? I don't see them mentioned in https://kubernetes.io/docs/setup/best-practices/certificates/

I'm not sure I understand. Requirements for doing what?

> But the code in this PR ignores the cert signer, which removes the entire point of using TLS in the first place.

This is test code. I'm not worried about some other pod impersonating controller manager or scheduler.

TLS is still useful for the production code that talks to those two pods. How to set that up properly is not something that I am familiar with.

johngmyers commented:
@pohly Requirements on cluster provisioners for how they need to provision controller manager and scheduler. This change broke several e2e tests for kubernetes/kops.

We believe #104061 might fix a lot of the breakage.

pohly commented:
> Requirements on cluster provisioners for how they need to provision controller manager and scheduler. This change broke several e2e tests for kubernetes/kops.

Can you elaborate which test broke and how?

The expectation now is that both controller manager and scheduler accept connections on localhost. Is that not the case on kops?

They also have to enable the secure port. But that was already the expectation before this PR here. TLS should be irrelevant for testing.

> We believe #104061 might fix a lot of the breakage.

That replaces `WaitForPodsReady`, which originally came from https://github.com/kubernetes/kubernetes/pull/89123/files i.e. not this PR. I've repeated the same mistake via copy-and-paste and it clearly needs to be fixed. I'm just trying to understand whether the concept behind this PR is okay.

johngmyers commented:
kOps, as of ten days ago, provisions server certs out of the Kubernetes General CA with SANs of `kube-controller-manager.kube-system.svc.cluster.local` and `kube-scheduler.kube-system.svc.cluster.local`, respectively. (replacing `cluster.local` with the cluster DNS name, if different)

It intentionally doesn't include `localhost` or `127.0.0.1` in the SAN, as such certs could be used to impersonate kube-apiserver to some clients.

An example of a broken test is https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/e2e-kops-warm-pool/1422330071519072256
Originally, both the controller-manager and scheduler metrics tests broke, but provisioning server certs as above fixed the scheduler test. So we think replacing `WaitForPodsReady` is likely to fix the other test. Since the scheduler test is now passing, it is unlikely that the lack of `localhost` in the SAN is going to be a problem, unlike what I earlier thought.

pohly commented:
> An example of a broken test is https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/e2e-kops-warm-pool/1422330071519072256

That error comes from https://github.com/kubernetes/kubernetes/blob/9ff3b7e744b34c099c1405d9add192adbef0b6b1/test/e2e/framework/metrics/metrics_grabber.go#L256-L260 and has nothing to do with the TLS certificates.

> Since the scheduler test is now passing, it is unlikely that the lack of localhost in the SAN is going to be a problem, unlike what I earlier thought.

So we agree that only this waiting for pods need to be fixed, but not any documentation or the port forwarding logic in this PR?
